File: scripts/tasks/makezip.py
import os
import shutil
import sys
import urllib.request
import zipfile

# IMPORTANT: This script must be run from the root of the repo!

base_url: str = "https://github.com/ZaparooProject/zaparoo.org/raw/refs/heads/main/docs/platforms/"
platform_docs: dict[str, str] = {
    "batocera": "batocera.md",
    "bazzite": "bazzite.mdx",
    "chimeraos": "chimeraos.mdx",
    "libreelec": "libreelec.mdx",
    "linux": "linux.mdx",
    "mac": "mac.mdx",
    "mister": "mister.md",
    "mistex": "mistex.md",
    "recalbox": "recalbox.mdx",
    "steamos": "steamos.md",
    "windows": "windows/index.md"
}
# files will be copied to the root of the zip
# dirs will copy the entire dir and preserve the structure
extra_items: dict[str, list[str]] = {
    "batocera": ["cmd/batocera/scripts"]
}


def strip_frontmatter(content: str) -> str:
    lines = content.splitlines()
    if lines[0] == "---":
        for i in range(1, len(lines)):
            if lines[i] == "---":
                return "\n".join(lines[i + 1:])
    return content


def download_doc(platform_id: str, to_dir: str):
    if platform_id not in platform_docs:
        raise ValueError(f"Platform '{platform_id}' not found in the platforms list.")

    file_name = platform_docs[platform_id]
    url = os.path.join(base_url, file_name)

    with urllib.request.urlopen(url) as response:
        content = response.read().decode("utf-8")

    if file_name.lower().endswith(".mdx"):
        content = strip_frontmatter(content)

    with open(os.path.join(to_dir, "README.txt"), "w", encoding="utf-8") as file:
        file.write(content.strip() + "\n")


if __name__ == "__main__":
    if len(sys.argv) < 5:
        raise ValueError("Usage: makezip.py <platform> <build_dir> <app_bin> <zip_name>")

    platform = sys.argv[1]
    build_dir = sys.argv[2]
    app_bin = sys.argv[3]
    zip_name = sys.argv[4]

    if not os.path.isdir(build_dir):
        raise NotADirectoryError(f"The specified directory '{build_dir}' does not exist.")

    license_path = os.path.join(build_dir, "LICENSE.txt")
    if not os.path.isfile(license_path):
        shutil.copy("LICENSE", license_path)

    app_path = os.path.join(build_dir, app_bin)
    if not os.path.isfile(app_path):
        raise FileNotFoundError(f"The specified binary file '{app_path}' does not exist.")

    zip_path = os.path.join(build_dir, zip_name)
    if os.path.isfile(zip_path):
        os.remove(zip_path)

    readme_path = os.path.join(build_dir, "README.txt")
    if not os.path.exists(readme_path):
        download_doc(platform, build_dir)

    with zipfile.ZipFile(zip_path, "w", zipfile.ZIP_DEFLATED) as dist:
        dist.write(app_path, arcname=os.path.basename(app_path))
        dist.write(license_path, arcname=os.path.basename(license_path))
        dist.write(readme_path, arcname=os.path.basename(readme_path))

        if platform in extra_items:
            for item in extra_items[platform]:
                if os.path.isfile(item):
                    # copy single files to the root of the zip
                    extra_file = os.path.join(build_dir, os.path.basename(item))
                    shutil.copy(item, build_dir)
                    dist.write(extra_file, arcname=os.path.basename(item))
                if os.path.isdir(item):
                    extra_dir = os.path.join(build_dir, os.path.basename(item))
                    if not os.path.exists(extra_dir):
                        os.makedirs(extra_dir)
                    shutil.copytree(item, extra_dir, dirs_exist_ok=True)
                    for root, dirs, files in os.walk(extra_dir):
                        for file in files:
                            path = str(os.path.join(root, file))
                            dist.write(path, arcname=os.path.join(os.path.relpath(root, build_dir), file))

File: scripts/tasks/utils/makezip.go
package main

import (
	"archive/zip"
	"fmt"
	"io"
	"net/http"
	"os"
	"path/filepath"
	"strings"
)

const baseURL = "https://github.com/ZaparooProject/zaparoo.org/raw/refs/heads/main/docs/platforms/"

var platformDocs = map[string]string{
	"batocera":  "batocera.md",
	"bazzite":   "bazzite.mdx",
	"chimeraos": "chimeraos.mdx",
	"libreelec": "libreelec.mdx",
	"linux":     "linux.mdx",
	"mac":       "mac.mdx",
	"mister":    "mister.md",
	"mistex":    "mistex.md",
	"recalbox":  "recalbox.mdx",
	"steamos":   "steamos.md",
	"windows":   "windows/index.md",
}

var extraItems = map[string][]string{
	"batocera": {"cmd/batocera/scripts"},
}

func stripFrontmatter(content string) string {
	lines := strings.Split(content, "\n")
	if len(lines) > 0 && lines[0] == "---" {
		for i := 1; i < len(lines); i++ {
			if lines[i] == "---" {
				return strings.Join(lines[i+1:], "\n")
			}
		}
	}
	return content
}

func downloadDoc(platformID, toDir string) error {
	fileName, ok := platformDocs[platformID]
	if !ok {
		return fmt.Errorf("platform '%s' not found in the platforms list", platformID)
	}

	url := baseURL + fileName
	resp, err := http.Get(url)
	if err != nil {
		return err
	}
	defer func(Body io.ReadCloser) {
		_ = Body.Close()
	}(resp.Body)

	content, err := io.ReadAll(resp.Body)
	if err != nil {
		return err
	}

	processedContent := string(content)
	if strings.HasSuffix(strings.ToLower(fileName), ".mdx") {
		processedContent = stripFrontmatter(processedContent)
	}

	return os.WriteFile(filepath.Join(toDir, "README.txt"), []byte(strings.TrimSpace(processedContent)+"\n"), 0644)
}

func main() {
	if len(os.Args) < 5 {
		fmt.Println("Usage: go run makezip.go <platform> <build_dir> <app_bin> <zip_name>")
		os.Exit(1)
	}

	platform := os.Args[1]
	buildDir := os.Args[2]
	appBin := os.Args[3]
	zipName := os.Args[4]

	if strings.HasPrefix(platform, "test") {
		os.Exit(0)
	}

	if _, err := os.Stat(buildDir); os.IsNotExist(err) {
		fmt.Printf("The specified directory '%s' does not exist\n", buildDir)
		os.Exit(1)
	}

	licensePath := filepath.Join(buildDir, "LICENSE.txt")
	if _, err := os.Stat(licensePath); os.IsNotExist(err) {
		input, err := os.ReadFile("LICENSE")
		if err != nil {
			fmt.Printf("Error reading LICENSE file: %v\n", err)
			os.Exit(1)
		}
		err = os.WriteFile(licensePath, input, 0644)
		if err != nil {
			fmt.Printf("Error copying LICENSE file: %v\n", err)
			os.Exit(1)
		}
	}

	appPath := filepath.Join(buildDir, appBin)
	if _, err := os.Stat(appPath); os.IsNotExist(err) {
		fmt.Printf("The specified binary file '%s' does not exist\n", appPath)
		os.Exit(1)
	}

	zipPath := filepath.Join(buildDir, zipName)
	_ = os.Remove(zipPath)

	readmePath := filepath.Join(buildDir, "README.txt")
	if _, err := os.Stat(readmePath); os.IsNotExist(err) {
		if err := downloadDoc(platform, buildDir); err != nil {
			fmt.Printf("Error downloading documentation: %v\n", err)
			os.Exit(1)
		}
	}

	zipFile, err := os.Create(zipPath)
	if err != nil {
		fmt.Printf("Error creating zip file: %v\n", err)
		os.Exit(1)
	}
	defer func(zipFile *os.File) {
		_ = zipFile.Close()
	}(zipFile)

	zipWriter := zip.NewWriter(zipFile)
	defer func(zipWriter *zip.Writer) {
		_ = zipWriter.Close()
	}(zipWriter)

	filesToAdd := []struct {
		path    string
		arcname string
	}{
		{appPath, filepath.Base(appPath)},
		{licensePath, filepath.Base(licensePath)},
		{readmePath, filepath.Base(readmePath)},
	}

	for _, file := range filesToAdd {
		err := addFileToZip(zipWriter, file.path, file.arcname)
		if err != nil {
			fmt.Printf("Error adding file to zip: %v\n", err)
			os.Exit(1)
		}
	}

	if items, ok := extraItems[platform]; ok {
		for _, item := range items {
			if info, err := os.Stat(item); err == nil {
				if info.IsDir() {
					err = addDirToZip(zipWriter, item, buildDir)
				} else {
					destPath := filepath.Join(buildDir, filepath.Base(item))
					if err := copyFile(item, destPath); err != nil {
						fmt.Printf("Error copying extra file: %v\n", err)
						os.Exit(1)
					}
					err = addFileToZip(zipWriter, destPath, filepath.Base(item))
				}
				if err != nil {
					fmt.Printf("Error adding extra item to zip: %v\n", err)
					os.Exit(1)
				}
			}
		}
	}
}

func addFileToZip(zipWriter *zip.Writer, filePath, arcname string) error {
	file, err := os.Open(filePath)
	if err != nil {
		return err
	}
	defer func(file *os.File) {
		_ = file.Close()
	}(file)

	info, err := file.Stat()
	if err != nil {
		return err
	}

	header, err := zip.FileInfoHeader(info)
	if err != nil {
		return err
	}
	header.Name = arcname
	header.Method = zip.Deflate

	writer, err := zipWriter.CreateHeader(header)
	if err != nil {
		return err
	}

	_, err = io.Copy(writer, file)
	return err
}

func addDirToZip(zipWriter *zip.Writer, dirPath string, buildDir string) error {
	return filepath.Walk(dirPath, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return err
		}

		if !info.IsDir() {
			relPath, err := filepath.Rel(dirPath, path)
			if err != nil {
				return err
			}

			destPath := filepath.Join(buildDir, filepath.Base(dirPath), relPath)
			if err := os.MkdirAll(filepath.Dir(destPath), 0755); err != nil {
				return err
			}

			if err := copyFile(path, destPath); err != nil {
				return err
			}

			return addFileToZip(zipWriter, destPath, filepath.Join(filepath.Base(dirPath), relPath))
		}
		return nil
	})
}

func copyFile(src, dst string) error {
	input, err := os.ReadFile(src)
	if err != nil {
		return err
	}
	return os.WriteFile(dst, input, 0644)
}
